error.go

```diff
@@ -302,6 +302,26 @@ func wrapIfContextError(err error) error {
 	return err
 }
 
+// wrapWithContextError wraps errors with CodeCanceled or CodeDeadlineExceeded
+// if the context is done. It leaves already-wrapped errors unchanged.
+func wrapWithContextError(ctx context.Context, err error) error {
+	if err == nil {
+		return nil
+	}
+	if _, ok := asError(err); ok {
+		return err
+	}
+	ctxErr := ctx.Err()
+	switch {
+	case errors.Is(ctxErr, context.Canceled):
+		return NewError(CodeCanceled, err)
+	case errors.Is(ctxErr, context.DeadlineExceeded):
+		return NewError(CodeDeadlineExceeded, err)
+	default:
+		return err
+	}
```

Comment:
This doesn't seem quite right. I would expect cases that look more like what is in `wrapIfContextError` (which is why I suggested combining them above).

This could make for confusing errors -- `err` could be unrelated to the cancellation, but we'll still stamp it as canceled or deadline exceeded code. For example, it could be a `*http.MaxBytesError`, which will result in a _very_ confusing message for one of these codes.

Reply:
Changed the ordering of the calls to avoid any possibility of wrapping a max bytes error with context codes.

Reply:
I'm not sure this really addresses the concern; it addresses only the one example I gave. In the face of other errors that might occur, unrelated to the context timing out or being cancelled -- it will unconditionally wrap an unrelated error with a context-related code. I guess it's okay, since the error will have to be I/O- or protocol-related (low-level protocol, like HTTP/2 framing or flow control errors) at this point (I think?), so maybe it's most likely related to the context error even if non-obviously. But it feels brittle.

Reply:
It was to avoid handling specific errors, like the http2 client disconnect, and instead use the context to capture a more suitable error code. Could be unrelated but most likely occurs due to an IO issues related to the context like you said.